### PR TITLE
added abstract dao, logging

### PR DESCRIPTION
**Pull request by yutsevich**

### CODE REVIEW

File: src/main/java/com/dev/cinema/service/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -20,8 +20,11 @@
 
     @Override
     public Order completeOrder(ShoppingCart shoppingCart) {
-        Order order = orderDao.add(new ArrayList<>(shoppingCart.getTickets()),
-                shoppingCart.getUser());
+        Order order = new Order();
+        order.setUser(shoppingCart.getUser());
+        order.setTickets(shoppingCart.getTickets());
```
Comment (MichaelShevchenko) [inline src/main/java/com/dev/cinema/service/impl/OrderServiceImpl.java]: вроде тут желательно реализовать как order.setTickets(new ArrayList<>(shoppingCart.getTickets())); из-за того кейса с очищением корзины вне зависимости от вызова метода clear()

**Review by MichaelShevchenko — APPROVED** [review]
Good job. We were asked to wrap login and registration into try-catch and to log error in catch block in main

File: src/main/java/com/dev/cinema/service/impl/OrderServiceImpl.java (hunk 1 above)
Comment (yutsevich) [reply src/main/java/com/dev/cinema/service/impl/OrderServiceImpl.java]: да. забыла пока меняла, спасибо

File: src/main/java/com/dev/cinema/Main.java [hunk 2]
```diff
@@ -16,22 +16,24 @@
...
+    private static final Logger logger = Logger.getLogger(Main.class);
 
     public static void main(String[] args) throws AuthenticationException {
         MovieService movieService = (MovieService) injector.getInstance(MovieService.class);
         Movie movie1 = new Movie();
         movie1.setTitle("Fast & Furious");
         movie1.setDescription("action");
         movieService.add(movie1);
-        movieService.getAll().forEach(System.out::println);
+        movieService.getAll().forEach(movie -> logger.info(movie.toString() + "is added"));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/Main.java]: no need to add " is added" to message, cause you have the same logger on dao layer. Here we can use method reference

File: src/main/java/com/dev/cinema/dao/AbstractDao.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.dev.cinema.dao.impl.CinemaHallDaoImpl;
+import com.dev.cinema.exceptions.DataProcessingException;
+import com.dev.cinema.util.HibernateUtil;
+import org.apache.log4j.Logger;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+public interface AbstractDao<T> {
+    Logger logger = Logger.getLogger(CinemaHallDaoImpl.class);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/dao/AbstractDao.java:11]: access modifier

File: src/main/java/com/dev/cinema/dao/AbstractDao.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    default T add(T entity) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(entity);
+            transaction.commit();
+            logger.info(entity.getClass().getSimpleName() + " "
+                    + entity.toString() + "is added");
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/dao/AbstractDao.java]:
```suggestion
                    + entity + "is added");
```
toString() method is called automatically

File: src/main/java/com/dev/cinema/dao/AbstractDao.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            return entity;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can not insert "
+                    + entity.getClass().getSimpleName()
+                    + entity.toString(), e);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/dev/cinema/dao/AbstractDao.java]: same

**Review by dmytro-hryhoruk — COMMENTED** [review]
Apply comments to all similar places

File: src/main/java/com/dev/cinema/dao/AbstractDao.java (hunk 3 above)
Comment (yutsevich) [reply src/main/java/com/dev/cinema/dao/AbstractDao.java:11]: private is not allowed, public is redundant, what should i use?

Comment (dmytro-hryhoruk) [reply src/main/java/com/dev/cinema/dao/AbstractDao.java:11]: Didn't notice that this is in interface. Then it's okay

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**